include/oneapi/dpl/experimental/kt/internal/esimd_radix_sort_kernels.h

```diff
@@ -0,0 +1,706 @@
+    constexpr _BinT __mask = __bin_count - 1;
+    constexpr ::std::uint32_t __hist_stride = sizeof(_HistT) * __bin_count;
+
+    constexpr ::std::uint32_t __reorder_slm_size = __data_per_work_item * sizeof(_KeyT) * __work_group_size;
+    constexpr ::std::uint32_t __bin_hist_slm_size = __hist_stride * __work_group_size;
+    constexpr ::std::uint32_t __incoming_offset_slm_size = (__bin_count + 1) * sizeof(_HistT);
+
+    // max SLM is 256 * 4 * 64 + 256 * 2 * 64 + 257*2, 97KB, when  __data_per_work_item = 256, __bin_count = 256
+    // to support 512 processing size, we can use all SLM as reorder buffer with cost of more barrier
+    __dpl_esimd::__ns::slm_init<::std::max(__reorder_slm_size, __bin_hist_slm_size + __incoming_offset_slm_size)>();
+
+    const ::std::uint32_t __local_tid = __idx.get_local_linear_id();
+
+    const ::std::uint32_t __slm_reorder_this_thread = __local_tid * __data_per_work_item * sizeof(_KeyT);
+    const ::std::uint32_t __slm_bin_hist_this_thread = __local_tid * __hist_stride;
+
+    __dpl_esimd::__ns::simd<_HistT, __bin_count> __bin_offset;
+    __dpl_esimd::__ns::simd<_DeviceAddrT, __data_per_work_item> __write_addr;
+    __dpl_esimd::__ns::simd<_KeyT, __data_per_work_item> __keys;
+    __dpl_esimd::__ns::simd<_BinT, __data_per_work_item> __bins;
+    __dpl_esimd::__ns::simd<_DeviceAddrT, __data_per_step> __lane_id(0, 1);
+
+    const _DeviceAddrT __io_offset = __data_per_work_item * __local_tid;
+
+    static_assert(__data_per_work_item % __data_per_step == 0);
+    static_assert(__bin_count % 128 == 0);
+    static_assert(__bin_count % 32 == 0);
+    _ONEDPL_PRAGMA_UNROLL
+    for (::std::uint32_t __s = 0; __s < __data_per_work_item; __s += __data_per_step)
+    {
+        __dpl_esimd::__ns::simd_mask<__data_per_step> __m = (__io_offset + __lane_id + __s) < __n;
+        __keys.template select<__data_per_step, 1>(__s) =
+            __dpl_esimd::__ns::merge(__dpl_esimd::__gather<_KeyT, __data_per_step>(__rng_data(__rng_pack.__keys_rng()), __lane_id, __io_offset + __s, __m),
+                  __dpl_esimd::__ns::simd<_KeyT, __data_per_step>(__sort_identity<_KeyT, __is_ascending>()), __m);
+    }
+
+    for (::std::uint32_t __stage = 0; __stage < __stage_count; __stage++)
+    {
+        __bins = __get_bucket<__mask>(__order_preserving_cast<__is_ascending>(__keys), __stage * __radix_bits);
+
+        __bin_offset = 0;
+        _ONEDPL_PRAGMA_UNROLL
+        for (::std::uint32_t __s = 0; __s < __data_per_work_item; __s += 1)
+        {
+            __write_addr[__s] = __bin_offset[__bins[__s]];
+            __bin_offset[__bins[__s]] += 1;
+        }
+
+        /*
+        first write to slm,
+        then do column scan by group, each thread to 32c*8r,
+        then last row do exclusive scan as incoming offset
+        then every thread add local sum with sum of previous group and incoming offset
+        */
+        {
```

Comment:
Does this bracket scope provide anything more than a conceptual count and scan section?  If not I would say we should delete it and just add more comments, or break it up into helper functions.  I think comments are sufficient.

Reply:
Extra scopes have been removed. Code blocks are now described.